### ISSUE TITLE
Support an output array parameter.

### ISSUE THREAD

**Issue by tbenthompson** [flatironinstitute/sparse_dot#9]
The MKL interface supports passing the output array for a matrix-vector product. (for example, see: https://scc.ustc.edu.cn/zlsc/tc4600/intel/2017.0.098/mkl/common/mklman_f/GUID-1FAA1369-97A7-4F13-87B2-3FE4425D876D.htm)

This is helpful when we want to avoid allocating a new array only to add it to an existing array. For example in the pseudocode snippet:

```
y = np.zeros(...)
for m in list_of_matrices:
    y += dot_product_mkl(m, x)
```

**Comment by asistradition** [comment]
That's no problem at all. The only limitation is that it'll have to be contiguous, so filling most array slices is right out.

**Comment by tbenthompson** [comment]
Yeah, that seems like a reasonable restriction. You could take a similar attitude to numpy.dot: https://numpy.org/doc/stable/reference/generated/numpy.dot.html

```
This is a performance feature. Therefore, if these conditions are not met, an exception is raised, instead of attempting to be flexible.
```

**Comment by asistradition** [comment]
I added `out` and `out_scalar` to `dot_product_mkl` such that `out := matrix_a * matrix_b + out_scalar * out` in 193e1182ef3e889684a71d30c2474ea7029b59ea. The `out` matrix needs to match what would have been produced anyway or it'll raise a `ValueError`.

I'm probably gonna write more tests for some of the edge conditions on this before I package it and put it onto pypi.

**Comment by tbenthompson** [comment]
Thanks! Makes sense that this would add a decent amount of new complexity and thus should be tested thoroughly. Really helpful that this package exists!

**Comment by asistradition** [comment]
Aight I think I've gotten all the subtle bugs. I've packaged this and released it as v0.6.0 - if there are any issues that come up, please let me know.
